File: movies.js
// movies.js
// Movie Catalog (Posters) на JavaScript (Node.js)

const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

// ANSI-цвета
const RESET = '\x1b[0m';
const BOLD = '\x1b[1m';
const CYAN = '\x1b[96m';
const GREEN = '\x1b[92m';
const YELLOW = '\x1b[93m';
const RED = '\x1b[91m';

function colorize(text, color) {
    return `${color}${text}${RESET}`;
}

class Movie {
    constructor(id, title, year, rating, poster, description = '') {
        this.id = id;
        this.title = title;
        this.year = year;
        this.rating = rating;
        this.poster = poster;
        this.description = description;
    }
}

class MovieCatalog {
    constructor(dataFile = 'movies.json') {
        this.dataFile = dataFile;
        this.movies = [];
        this.nextId = 1;
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(this.dataFile)) {
                const data = JSON.parse(fs.readFileSync(this.dataFile, 'utf-8'));
                this.movies = data.movies.map(m => new Movie(m.id, m.title, m.year, m.rating, m.poster, m.description));
                this.nextId = data.next_id || 1;
            }
        } catch (e) {
            this.movies = [];
            this.nextId = 1;
        }
    }

    save() {
        const data = {
            movies: this.movies.map(m => ({ id: m.id, title: m.title, year: m.year, rating: m.rating, poster: m.poster, description: m.description })),
            next_id: this.nextId
        };
        fs.writeFileSync(this.dataFile, JSON.stringify(data, null, 2), 'utf-8');
    }

    add(title, year, rating, poster, description = '') {
        const movie = new Movie(this.nextId, title, year, rating, poster, description);
        this.movies.push(movie);
        this.nextId++;
        this.save();
        return movie;
    }

    remove(id) {
        const idx = this.movies.findIndex(m => m.id === id);
        if (idx === -1) return false;
        this.movies.splice(idx, 1);
        this.save();
        return true;
    }

    get(id) {
        return this.movies.find(m => m.id === id) || null;
    }

    search(query) {
        const q = query.toLowerCase();
        return this.movies.filter(m => m.title.toLowerCase().includes(q));
    }

    list() {
        return this.movies;
    }

    openPoster(id) {
        const movie = this.get(id);
        if (!movie) return false;
        const posterPath = movie.poster;
        if (!fs.existsSync(posterPath)) {
            console.log(colorize(`Постер не найден: ${posterPath}`, RED));
            return false;
        }
        const cmd = process.platform === 'darwin' ? 'open' :
                    process.platform === 'win32' ? 'start' : 'xdg-open';
        exec(`${cmd} "${posterPath}"`);
        return true;
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0 || args[0] === 'help') {
        console.log(`Использование: node movies.js <команда> [опции]
  add       --title <title> --year <year> --rating <rating> --poster <poster> [--description <desc>]
  remove    --id <id>
  list
  search    --query <query>
  info      --id <id>
  open      --id <id>
  help`);
        process.exit(0);
    }

    const command = args[0];
    const options = {};
    for (let i = 1; i < args.length; i++) {
        if (args[i].startsWith('--')) {
            const key = args[i].slice(2);
            const value = args[++i];
            options[key] = value;
        }
    }

    const dataFile = options.data || 'movies.json';
    const catalog = new MovieCatalog(dataFile);

    switch (command) {
        case 'add': {
            if (!options.title || !options.year || !options.rating || !options.poster) {
                console.error('Ошибка: требуются --title, --year, --rating, --poster');
                process.exit(1);
            }
            const movie = catalog.add(options.title, parseInt(options.year), parseFloat(options.rating), options.poster, options.description || '');
            console.log(colorize(`Фильм добавлен: ID ${movie.id} - ${movie.title}`, GREEN));
            break;
        }
        case 'remove': {
            if (!options.id) {
                console.error('Ошибка: укажите --id');
                process.exit(1);
            }
            const id = parseInt(options.id);
            if (catalog.remove(id)) {
                console.log(colorize(`Фильм с ID ${id} удалён`, GREEN));
            } else {
                console.log(colorize(`Фильм с ID ${id} не найден`, RED));
            }
            break;
        }
        case 'list': {
            const movies = catalog.list();
            if (movies.length === 0) {
                console.log('Нет фильмов.');
            } else {
                for (const m of movies) {
                    console.log(`${colorize(String(m.id), CYAN)} | ${colorize(m.title, BOLD)} | ${m.year} | ${colorize(String(m.rating), YELLOW)} | ${m.poster}`);
                }
            }
            break;
        }
        case 'search': {
            if (!options.query) {
                console.error('Ошибка: укажите --query');
                process.exit(1);
            }
            const results = catalog.search(options.query);
            if (results.length === 0) {
                console.log('Ничего не найдено.');
            } else {
                for (const m of results) {
                    console.log(`${colorize(String(m.id), CYAN)} | ${colorize(m.title, BOLD)} | ${m.year} | ${colorize(String(m.rating), YELLOW)} | ${m.poster}`);
                }
            }
            break;
        }
        case 'info': {
            if (!options.id) {
                console.error('Ошибка: укажите --id');
                process.exit(1);
            }
            const id = parseInt(options.id);
            const movie = catalog.get(id);
            if (!movie) {
                console.log(colorize(`Фильм с ID ${id} не найден`, RED));
            } else {
                console.log(colorize(`ID: ${movie.id}`, CYAN));
                console.log(colorize(`Название: ${movie.title}`, BOLD));
                console.log(`Год: ${movie.year}`);
                console.log(`Рейтинг: ${colorize(String(movie.rating), YELLOW)}`);
                console.log(`Постер: ${movie.poster}`);
                console.log(`Описание: ${movie.description}`);
            }
            break;
        }
        case 'open': {
            if (!options.id) {
                console.error('Ошибка: укажите --id');
                process.exit(1);
            }
            const id = parseInt(options.id);
            if (catalog.openPoster(id)) {
                console.log(colorize(`Постер фильма ID ${id} открыт`, GREEN));
            } else {
                console.log(colorize(`Не удалось открыть постер для ID ${id}`, RED));
            }
            break;
        }
        default:
            console.log('Неизвестная команда.');
    }
}

main();

File: movies.py
# movies.py
# Movie Catalog (Posters) на Python

import sys
import os
import json
import argparse
import subprocess
from datetime import datetime
from typing import List, Dict, Optional

# ANSI-цвета для красивого вывода
RESET = "\033[0m"
BOLD = "\033[1m"
CYAN = "\033[96m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"

def colorize(text, color):
    return f"{color}{text}{RESET}"

class Movie:
    def __init__(self, id: int, title: str, year: int, rating: float, poster: str, description: str = ""):
        self.id = id
        self.title = title
        self.year = year
        self.rating = rating
        self.poster = poster
        self.description = description

    def to_dict(self):
        return {
            "id": self.id,
            "title": self.title,
            "year": self.year,
            "rating": self.rating,
            "poster": self.poster,
            "description": self.description
        }

    @classmethod
    def from_dict(cls, data):
        return cls(data["id"], data["title"], data["year"], data["rating"], data["poster"], data.get("description", ""))

class MovieCatalog:
    def __init__(self, data_file="movies.json"):
        self.data_file = data_file
        self.movies: List[Movie] = []
        self.next_id = 1
        self.load()

    def load(self):
        if os.path.exists(self.data_file):
            try:
                with open(self.data_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.movies = [Movie.from_dict(m) for m in data.get("movies", [])]
                    self.next_id = data.get("next_id", 1)
            except:
                self.movies = []
                self.next_id = 1
        else:
            self.movies = []
            self.next_id = 1

    def save(self):
        data = {
            "movies": [m.to_dict() for m in self.movies],
            "next_id": self.next_id
        }
        with open(self.data_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)

    def add(self, title: str, year: int, rating: float, poster: str, description: str = "") -> Movie:
        movie = Movie(self.next_id, title, year, rating, poster, description)
        self.movies.append(movie)
        self.next_id += 1
        self.save()
        return movie

    def remove(self, movie_id: int) -> bool:
        for i, m in enumerate(self.movies):
            if m.id == movie_id:
                del self.movies[i]
                self.save()
                return True
        return False

    def get(self, movie_id: int) -> Optional[Movie]:
        for m in self.movies:
            if m.id == movie_id:
                return m
        return None

    def search(self, query: str) -> List[Movie]:
        q = query.lower()
        return [m for m in self.movies if q in m.title.lower()]

    def list_all(self) -> List[Movie]:
        return self.movies

    def open_poster(self, movie_id: int) -> bool:
        movie = self.get(movie_id)
        if not movie:
            return False
        poster_path = movie.poster
        if not os.path.exists(poster_path):
            print(colorize(f"Постер не найден: {poster_path}", RED))
            return False
        # Открываем во внешнем просмотрщике
        if sys.platform == 'darwin':
            subprocess.run(['open', poster_path])
        elif sys.platform == 'win32':
            subprocess.run(['start', poster_path], shell=True)
        else:
            subprocess.run(['xdg-open', poster_path])
        return True

def main():
    parser = argparse.ArgumentParser(description="Movie Catalog (Posters)")
    parser.add_argument("command", choices=["add", "remove", "list", "search", "info", "open", "help"],
                        help="Команда")
    parser.add_argument("--title", help="Название фильма")
    parser.add_argument("--year", type=int, help="Год выпуска")
    parser.add_argument("--rating", type=float, help="Рейтинг")
    parser.add_argument("--poster", help="Путь к постеру")
    parser.add_argument("--description", help="Описание")
    parser.add_argument("--id", type=int, help="ID фильма")
    parser.add_argument("--query", help="Поисковый запрос")
    parser.add_argument("--data", default="movies.json", help="Файл данных")
    args = parser.parse_args()

    catalog = MovieCatalog(args.data)

    if args.command == "help":
        print(__doc__)
        sys.exit(0)

    elif args.command == "add":
        if not args.title or not args.year or not args.rating or not args.poster:
            print("Ошибка: требуются --title, --year, --rating, --poster")
            sys.exit(1)
        movie = catalog.add(args.title, args.year, args.rating, args.poster, args.description or "")
        print(colorize(f"Фильм добавлен: ID {movie.id} - {movie.title}", GREEN))

    elif args.command == "remove":
        if args.id is None:
            print("Ошибка: укажите --id")
            sys.exit(1)
        if catalog.remove(args.id):
            print(colorize(f"Фильм с ID {args.id} удалён", GREEN))
        else:
            print(colorize(f"Фильм с ID {args.id} не найден", RED))

    elif args.command == "list":
        movies = catalog.list_all()
        if not movies:
            print("Нет фильмов.")
        else:
            for m in movies:
                print(f"{colorize(str(m.id), CYAN)} | {colorize(m.title, BOLD)} | {m.year} | {colorize(str(m.rating), YELLOW)} | {m.poster}")

    elif args.command == "search":
        if not args.query:
            print("Ошибка: укажите --query")
            sys.exit(1)
        results = catalog.search(args.query)
        if not results:
            print("Ничего не найдено.")
        else:
            for m in results:
                print(f"{colorize(str(m.id), CYAN)} | {colorize(m.title, BOLD)} | {m.year} | {colorize(str(m.rating), YELLOW)} | {m.poster}")

    elif args.command == "info":
        if args.id is None:
            print("Ошибка: укажите --id")
            sys.exit(1)
        movie = catalog.get(args.id)
        if not movie:
            print(colorize(f"Фильм с ID {args.id} не найден", RED))
        else:
            print(colorize(f"ID: {movie.id}", CYAN))
            print(colorize(f"Название: {movie.title}", BOLD))
            print(f"Год: {movie.year}")
            print(f"Рейтинг: {colorize(str(movie.rating), YELLOW)}")
            print(f"Постер: {movie.poster}")
            print(f"Описание: {movie.description}")

    elif args.command == "open":
        if args.id is None:
            print("Ошибка: укажите --id")
            sys.exit(1)
        if catalog.open_poster(args.id):
            print(colorize(f"Постер фильма ID {args.id} открыт", GREEN))
        else:
            print(colorize(f"Не удалось открыть постер для ID {args.id}", RED))

if __name__ == "__main__":
    main()
